snapcraft/cli/lifecycle.py

```diff
@@ -62,11 +63,15 @@ def _execute(
         with build_provider_class(project=project, echoer=echo) as instance:
             instance.mount_project()
             try:
-                if pack_project:
+                if pack_project and shell:
```

Comment:
I would've structured this a little differently, to be easier to understand:
```python
if shell:
    # do all the work until just before the wanted step
    if pack_project: ... else ... yadda
else:
   # ... current code
```
and then have the shell / shell_after checking outside the try/except as noted before :-)